src/llm/io_processing/qwen3coder/qwen3coder_tool_parser.hpp

```diff
@@ -0,0 +1,165 @@
+    /*
+     * Return all tool calls found in agreggated content so far
+     * that were not returned before
+     */
+    std::optional<ToolCalls> parseChunk(const std::string& chunk);
+    std::optional<std::string> getCurrentFunctionName() const;
+    Status removeToolCallsFromContentIfNeeded(std::string& outContent);
+    State getCurrentState() const {
+        return this->currentState;
+    }
+    size_t getLastProcessedPosition() const {
+        return this->lastProcessedPosition;
+    }
+
+private:
+    const ToolsParameterTypeMap_t& toolsParametersTypeMap;
+    const bool removeNewlineAroundParameters = true;
+    State currentState = State::Content;
+    Functool currentFunction;
+    std::string currentParameterName;
+    std::string streamContent;  // content accumulated from stream chunks
+    // current position in content
+    size_t lastProcessedPosition{0};
+    // members required for unary for removing tool calls from content
+    std::stack<size_t> toolsBeginStack;
+    std::stack<size_t> toolsEndStack;
+    /*
+     * process streamcontent from lastProcessedPosition until change of state
+     * return true if state changed, false otherwise
+     * false means no more state changes possible with current content
+     */
+    bool stepUntilStateChange(ToolCalls& toolCalls);
+};
+
+class Qwen3CoderToolParser : public BaseOutputParser {
+public:
+    static const std::string toolsStartTag;
+    static const std::string toolsEndTag;
+    static const std::string toolPrefixTag;
+    static const std::string toolEndTag;
+    static const std::string parameterPrefixTag;
+    static const std::string parameterEndTag;
+    static const std::string tagEnd;
+
+private:
+    const ToolsSchemas_t& toolSchemas;  // we need to keep reference as this is not filled in OpenAIApiHandler during ToolParser creation, NOTE that its const here but it can change outside
+    ToolsParameterTypeMap_t toolsParametersTypes;
+    bool filledParametersTypesMap{false};
+    // for streaming parsing we need to keep parser as a member
+    Qwen3CoderToolParserImpl streamParser;
+    int toolCallIndex{-1};
+    ToolCalls currentToolCalls;
+    rapidjson::Document currentJson;
+    std::set<int> returnedFirstDeltas;
+    std::set<int> returnedCompleteDeltas;
+    // streaming off
+
+public:
+    Qwen3CoderToolParser() = delete;
+    explicit Qwen3CoderToolParser(ov::genai::Tokenizer& tokenizer, const ToolsSchemas_t& toolSchemas);
+
+    void parse(ParsedOutput& parsedOutput, const std::vector<int64_t>& generatedTokens) override;
+    std::optional<rapidjson::Document> parseChunk(const std::string& chunk, ov::genai::GenerationFinishReason finishReason) override;
+    const std::string& getParsingStartTag() const override {
+        return toolsStartTag;
+    }
+    const std::unordered_set<std::string>& getSpecialParsingStartTags() const override {
+        static const std::unordered_set<std::string> specialParsingStartTags = {toolsStartTag};
```

Comment:
This should probably be empty. If qwen3 coder has only one way (one tag) to start tool parsing, it should not be repeated in special tags.